### PR TITLE
558 implement a base class for scripting models

### LINKED ISSUE #558: Implement a base class for scripting models
**Is your feature request related to a problem? Please describe.**
For inference and deployment of models in production, I would like to create serializable models. Such models allow us to run inference without the code to build the architecture; furthermore, input preprocessing functions can be scripted with the models hence minimizing input drift when using models.

**Describe the solution you'd like**
A base class for scripting models during training; the function call is optional. 

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by valhassan**
Thank you for submitting your **PR** and helping us with **GDL**.
Please read the template below, fill it out as appropriate, and make additional changes to your code as needed. 
Please feel free to submit your PR even if it doesn't satisfy all of the requirements below.
If you need assistance or review from a maintainer, add the label __Status: Help Needed__ or __Status: Review Needed__ respectively.
After review, a maintainer will add the label __Status: Revision Needed__ if further work is required for the PR to be merged.

# Description
Please include a summary of the change and which issue is resolved (if applicable). 
Please also include relevant motivation and context. 
List any dependencies that are required for this change.
If it fixes an open issue, please link to the issue here.
Fixes #558 

## Type of change
What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other 

## How Has This Been Tested?
If the added changes are already cover by tests, skip that section.
Otherwise, please describe tests that you added to the pytest codebase (if applicable).

## Screenshots (if applicable):

# Checklist:
If you're unsure about any of these, don't hesitate to ask. We're here to help! 
- [x] My code follows the code style of this project
- [x] I have performed a self-review of my own code
- [x] My PR has a descriptive title
- [x] My code follows PEP8 (most of it)
- [x] My code have the proper documentation in the code
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] My changes generate no new errors
- [x] My code is covered by the existing tests (if applicable)
- [ ] My code includes tests that show that my feature works
- [x] My PR passes Travis CI tests

_If your PR does not fulfill all of the requirements in the checklist above, that's OK!_ If you need help, @-mention a maintainer and/or add the __Status: Help Needed__ label.

### CODE REVIEW

File: config/gdl_config_template.yaml [hunk 1]
```diff
@@ -34,7 +34,7 @@ general:
   raw_data_dir: data
   raw_data_csv: tests/tiling/tiling_segmentation_binary_ci.csv
   tiling_data_dir: ${general.raw_data_dir}/patches # where the patches will be saved
-  save_weights_dir: saved_model/${general.project_name}
```
Comment (CharlesAuthier) [inline config/gdl_config_template.yaml:37]: why changing the template?

**Review by CharlesAuthier — APPROVED** [review]

File: config/gdl_config_template.yaml (hunk 1 above)
Comment (valhassan) [reply config/gdl_config_template.yaml]: Change reverted beause it fails tests!

**Review by CharlesAuthier — APPROVED** [review]

File: utils/script_model.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import torch
+
+class ScriptModel(torch.nn.Module):
```
Comment (remtav) [inline utils/script_model.py:3]: Hey! Very curious to start using scripted models. Thanks for this addition. Could a test for the __init__ and forward() to validate these methods act as expected be relevant?

Comment (valhassan) [reply utils/script_model.py:3]: yes, I added tests.

**Review by CharlesAuthier — APPROVED** [review]

File: train_segmentation.py [hunk 3]
```diff
@@ -792,6 +794,19 @@ def train(cfg: DictConfig) -> None:
...
+        model_to_script = ScriptModel(model,
+                                      device=device,
+                                      input_shape=(1, num_bands, patches_size, patches_size),
+                                      mean=mean,
+                                      std=std,
+                                      min=scale[0],
+                                      max=scale[1])
+        
+        scripted_model = torch.jit.script(model_to_script)
+        scripted_model.save(output_path.joinpath('scripted_model.pt'))
```
[the remark this answers is not in the document]
Comment (valhassan) [reply train_segmentation.py:809]: The scripting logic is after all training and validation loops are completed. This means scripting and saving the resulting model is done just once. Also, the whole scripting logic is enabled by choice with a flag; hence you can run it without training your model. 

On the generic name, I did not give it much thought; any suggestions we can explore? One solution is to utilize the naming convention for the raw weights that's saved at latest best validation loss.

[the remark this answers is not in the document]
Comment (LucaRom) [reply train_segmentation.py:809]: Thanks for the clarifications. The naming is only important if it will be shared outside of the output_path. I think we can look at it later if we see it necessary.

**Review by LucaRom — APPROVED** [review]
